Fix `Table.scan` to enable case sensitive argument

jiakai-li opened the pull request:
**This pull request fixes below issue:**
- #1421 

The change modified the `DataScan._build_partition_projection` method to pass `case_sensitive` argument when calling `inclusive_projection`. So that `Table.scan` method respects this configuration. Also added related tests.

sungwy commented:
Hi @jiakai-li - thank you for reporting this issue and putting up this PR!

The behavior looks right to me, but I think it would be worth thinking through this behavior change holistically. In `_DeleteFiles` we also build partition projection using `inclusive_projection` where we only support `case_sensitive` matching. https://github.com/apache/iceberg-python/blob/ede363bd63640f3db6c3b3d6904a5765dafd2600/pyiceberg/table/update/snapshot.py#L340-L344

If we introduce `case_sensitive=False` as a behavior that can be controlled by the user on DataScan for `inclusive_projection`, I think its worth also thinking if we want to introduce a discrepancy in our behaviors based on API, or if we should be keeping a consistent behavior (one way or another).

Tagging @Fokko for review

Fokko commented:
First of all, thanks @jiakai-li for raising this PR, and thanks @sungwy for raising that issue. And I agree, I think we should also be able to control case-sensitivity when doing deletes (defaulting to `true` as this is the current behavior).

sungwy commented:
Yes, I think updating this PR to include the changes for both makes sense @jiakai-li 👍 

Thank you again for tackling this issue!

kevinjqliu commented on tests/table/test_init.py:
```diff
@@ -310,6 +310,19 @@ def test_table_scan_row_filter(table_v2: Table) -> None:
     assert scan.filter(EqualTo("x", 10)).filter(In("y", (10, 11))).row_filter == And(EqualTo("x", 10), In("y", (10, 11)))
 
 
+def test_table_scan_partition_filters_case_sensitive(table_v2: Table) -> None:
```
nit: the goal of this test is to show that case sensitive scan works. i think a more readable test can be something like
1. assert data exists for `row_filter=EqualTo("X", 10)`
2. assert case sensitive scan for `row_filter=EqualTo("X", 10)` returns the same data
3. assert case insensitive scan for `row_filter=EqualTo("X", 10)` and `row_filter=EqualTo("x", 10)` also returns the same data

jiakai-li replied:
Thanks very much @kevinjqliu , I've updated the test case for scan and the other operations as well. Hope it looks better now?

kevinjqliu reviewed (commented):
Thanks for the PR! Left a comment on making the tests more readable

jiakai-li commented:
Hey guys, thanks a lot for your kind guidance and great suggestions. I've updated the PR to:
- Enable `Table.delete` and `Table.overwrite` operations to control case-sensitivity
- Updated test cases to make it more readable

Please let me know if there is other changes you would like me to make, thanks.

kevinjqliu commented on tests/integration/test_deletes.py:
```diff
@@ -770,3 +789,67 @@ def test_delete_after_partition_evolution_from_partitioned(session_catalog: Rest
+def test_delete_with_filter_case_sensitive(test_table: Table) -> None:
+    assert {"idx": 2, "value": "b"} in test_table.scan().to_arrow().to_pylist()
+
+    with pytest.raises(ValueError) as e:
```
nit: also test default test sensitivity `test_table.delete("Idx == 2")`

kevinjqliu replied:
same for other tests

kevinjqliu commented on tests/integration/test_deletes.py:
```diff
@@ -770,3 +789,67 @@ def test_delete_after_partition_evolution_from_partitioned(session_catalog: Rest
+
+@pytest.mark.integration
+def test_delete_with_filter_case_sensitive(test_table: Table) -> None:
+    assert {"idx": 2, "value": "b"} in test_table.scan().to_arrow().to_pylist()
```
nit: set 
```
expected_value = {"idx": 2, "value": "b"}
```
and reuse the variable

kevinjqliu reviewed (commented):
added a few nit comments, generally LGTM! Thanks!

jiakai-li commented:
> added a few nit comments, generally LGTM! Thanks!

Thank you @kevinjqliu ! I've updated the test cases and resolve the comments. Please let me know if you have further comments.

Fokko commented on pyiceberg/table/__init__.py:
```diff
@@ -987,6 +993,7 @@ def overwrite(
         self,
         df: pa.Table,
         overwrite_filter: Union[BooleanExpression, str] = ALWAYS_TRUE,
+        case_sensitive: bool = True,
```
Same here, let's add this to the end to avoid breaking folks who use positional arguments to pass in properties.

Fokko commented on pyiceberg/table/__init__.py:
```diff
@@ -268,12 +268,10 @@ def _apply(self, updates: Tuple[TableUpdate, ...], requirements: Tuple[TableRequ
         return self
 
-    def _scan(self, row_filter: Union[str, BooleanExpression] = ALWAYS_TRUE) -> DataScan:
+    def _scan(self, row_filter: Union[str, BooleanExpression] = ALWAYS_TRUE, case_sensitive: bool = True) -> DataScan:
```
I would leave this argument out since the `TableScan` is more like a builder pattern: `.scan().with_case_sensitive()`. Also, it doesn't allow to pass in a filter, columns, etc.

jiakai-li replied:
Thanks Fokko :-) I've updated the PR to adjust the `case_sensitive` argument position to the end and closed the other comments. But I might need some more guidance regarding this one.

My understanding is this `Transaction._scan` method does not relate to the `TableScan`. It seems a helper method only used by the `Transaction.delete` method, which now takes the new `case_sensitive` and use it to plan files. I can modify the `Transaction.delete` method to use `DataScan` directly in order to pass `case_sensitive`, but that would then make this `Transaction._scan` redundant. Should I do it this way?

Fokko replied:
Hey @jiakai-li Sorry for my ambiguous message, let me clarify. The `._scan()` returns a `DataScan` that accepts very few arguments because it follows a builder pattern. With also adding the arguments, we invalidate that pattern. Instead, below it should use `._scan().with_case_sensitive(case_sensitive)`. However, it looks like this ship has sailed already, and let's not block it since this is just a private method.

jiakai-li replied:
Thanks for the explaination @Fokko , now I get it :-)

Fokko commented on pyiceberg/table/__init__.py:
```diff
@@ -421,6 +419,7 @@ def overwrite(
         self,
         df: pa.Table,
         overwrite_filter: Union[BooleanExpression, str] = ALWAYS_TRUE,
+        case_sensitive: bool = True,
```
Also, this one should go to the very end. I'll stop commenting on this :)

Fokko reviewed (commented):
Thanks again @jiakai-li this looks very good. I've left some small comments, but in general this looks good 👍

Fokko reviewed (approved):
This looks great, thanks @jiakai-li for working on this, and thanks @sungwy for the sharp eye 👍
